src/main/scala/firrtl/passes/clocklist/ClockList.scala

```diff
@@ -0,0 +1,72 @@
+// See license file for details
+
+package firrtl.passes
+package clocklist
+
+import firrtl._
+import firrtl.ir._
+import Annotations._
+import Utils.error
+import java.io.{File, CharArrayWriter, PrintWriter, Writer}
+import wiring.WiringUtils.{getChildrenMap, countInstances, ChildrenMap, getLineage}
+import wiring.Lineage
+import ClockListUtils._
+import Utils._
+import memlib.AnalysisUtils._
+import memlib._
+import Mappers._
+
+/** Starting with a top module, determine the clock origins of each child instance.
+ *  Write the result to writer.
+ */
+class ClockList(top: String, writer: Writer) extends Pass {
+  def name = this.getClass.getSimpleName
+  def run(c: Circuit): Circuit = {
+    // Build useful datastructures
+    val childrenMap = getChildrenMap(c)
+    val moduleMap = c.modules.foldLeft(Map[String, DefModule]())((map, m) => map + (m.name -> m))
+    val lineages = getLineage(childrenMap, top)
+    val outputBuffer = new CharArrayWriter
+
+    // === Checks ===
+    // TODO(izraelevitz): Check all registers/memories use "clock" clock port
+    // ==============
+    
+    // Clock sources must be blackbox outputs and top's clock
+    val partialSourceList = getSourceList(moduleMap)(lineages)
+    val sourceList = partialSourceList ++ moduleMap(top).ports.collect{ case Port(i, n, Input, ClockType) => n }
+    writer.append(s"Sourcelist: $sourceList \n")
+
+    // Remove everything from the circuit, unless it has a clock type
+    // This simplifies the circuit drastically so InlineInstances doesn't take forever.
+    val onlyClockCircuit = RemoveAllButClocks.run(c)
+    
+    // Inline the clock-only circuit up to the specified top module
+    val modulesToInline = (c.modules.collect { case Module(_, n, _, _) if n != top => ModuleName(n, CircuitName(c.main)) }).toSet
+    val inlineTransform = new InlineInstances
+    val inlinedCircuit = inlineTransform.run(onlyClockCircuit, modulesToInline, Set()).circuit
+    val topModule = inlinedCircuit.modules.collectFirst { case m if m.name == top => m }.get
```

Comment:
I think this would be better as a find:
```scala
inlinedCircuit.modules.find(_.name == top).get
```
And possibly it should be `.getOrElse(throwInternalError)`